staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -147,34 +151,100 @@ func (r *withRetry) NextRetry(req *http.Request, resp *http.Response, err error,
-	// Ensure the response body is fully read and closed before
-	// we reconnect, so that we reuse the same TCP connection.
+// prepareForNextRetry is responsible for carrying out operations that need
+// to be completed before the next retry is initiated:
+// - if the request context is already canceled there is no need to
+//   retry, the function will return ctx.Err().
+// - we need to seek to the beginning of the request body before we
+//   initiate the next retry, the function should return an error if
+//   it fails to do so.
+func (r *withRetry) prepareForNextRetry(ctx context.Context, request *Request) error {
 	if ctx.Err() != nil {
 		return ctx.Err()
 	}
 
-	if seeker, ok := body.(io.Seeker); ok && body != nil {
+	// Ensure the response body is fully read and closed before
+	// we reconnect, so that we reuse the same TCP connection.
+	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
 		if _, err := seeker.Seek(0, 0); err != nil {
-			return fmt.Errorf("can't Seek() back to beginning of body for %T", r)
+			return fmt.Errorf("can't Seek() back to beginning of body for %T", request)
 		}
 	}
 
-	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", retryAfter.Wait, retryAfter.Attempt, url)
-	if backoff != nil {
-		backoff.Sleep(retryAfter.Wait)
+	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", r.retryAfter.Wait, r.retryAfter.Attempt, request.URL().String())
+	return nil
+}
+
+func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	url := request.URL()
+
+	// r.retryAfter represents the retry after parameters calculated
+	// from the (response, err) tuple from the last attempt, so 'Before'
+	// can apply these retry after parameters prior to the next attempt.
+	// 'r.retryAfter == nil' indicates that this is the very first attempt.
+	if r.retryAfter == nil {
+		// we do a backoff sleep before the first attempt is made,
+		// (preserving current behavior).
+		request.backoff.Sleep(request.backoff.CalculateBackoff(url))
+		return nil
+	}
+
+	// if we are here, we have made attempt(s) al least once before.
+	if request.backoff != nil {
+		// TODO(tkashem) with default set to use exponential backoff
+		//  we can merge these two sleeps:
+		//  BackOffManager.Sleep(max(backoffManager.CalculateBackoff(), retryAfter))
+		//  see https://github.com/kubernetes/kubernetes/issues/108302
+		request.backoff.Sleep(r.retryAfter.Wait)
+		request.backoff.Sleep(request.backoff.CalculateBackoff(url))
 	}
+
+	// We are retrying the request that we already send to
+	// apiserver at least once before. This request should
+	// also be throttled with the client-internal rate limiter.
+	if err := request.tryThrottleWithInfo(ctx, r.retryAfter.Reason); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (r *withRetry) After(ctx context.Context, request *Request, resp *http.Response, err error) {
```

Comment:
Why are we updating backoff here?  Why can't this be in Before?

Reply:
we update the backoff after we get an answer tuple `(response, err)` from the server.  If we store the answer tuple maybe we can get rid of `Before`. I can look into it when I do the follow up refactor.